### ISSUE TITLE
proposal: small change to Axis to reveal "natural" dimensions

### ISSUE THREAD

**Issue by tr8dr** [oxyplot/oxyplot#383]
## Problem
I have a series of PlotModels (Views) where I want to align 1 axis.  For example I may have a series of plots aligned vertically with the same X axis, but with different Y axes.

A problem to solve is that the render-width of each of these Y axes may be different (and in fact may change as a given plot is panned).   I can set __<model>.PlotMargins__ left margin to a fixed value, but this may clip the rendering for an axis.  The __ActualPlotMargin__ property will not adjust to the natural sizing if the __PlotMargin__ dimension is overridden.  Hence, I have a circular problem in adjusting the margins across panes in that if adjust cannot remeasure the natural sizing.

The rendering calls:
```csharp
 axis.Measure(IRenderContext);
``` 
on each axis and adjust to the max extent across the set.   This information is thrown-away if the PlotMargin dimension is set and not accessible at the application level.

## Proposal
Add an __Update(IRenderContext)__ method to Axis that either calls Measure(rc) or replaces it.  The method updates a property:  

```csharp
Thickness ActualDimensions { get; protected set; }
```
The internal renderer and application, alike, may now read this property.

**Comment by objorke** [comment]
I don't think the layout "engine" should be exposed. To make it possible to maintain/refactor this engine, we should not expose any of the internal steps. Can we find a different solution to this issue?

Adding a property that contains the actual location and size of the axis sounds good. But it should use start with "Actual" for consistency.

**Comment by tr8dr** [comment]
The layout methods can be protected.  I don't need to see these as a user of the library.   What I do need is a property which exposes the "desired" rendering width / height of the axis.   The desired height/width is distinct from the Actual height or width, as actual may reflect a user size override.

**Comment by objorke** [comment]
I think I get it now. `DesiredSize` as in WPF [`UIElement.DesiredSize`](https://msdn.microsoft.com/en-us/library/system.windows.uielement.desiredsize%28v=vs.110%29.aspx) is a better name than `NaturalSize` in my opinion.

**Comment by tr8dr** [comment]
I agree: DesiredSize sounds good.  Ideally this makes its way across other UI elements.   How would you propose unit testing?

**Comment by objorke** [comment]
The WPF example in 6b10059 should be updated to demonstrate the functionality described in this issue.

**Comment by tr8dr** [comment]
I can take a look at this when time permits (I would need to install a VM to work through it, as I run OS X exclusively).   With regard to making this work, there is a feature that would be useful:

I've managed to get the dynamic alignment to work by both watching Axis.AxisChanged events and also creating my own "OnInitialRender" event.   Without panning or otherwise causing the axes to change, the alignment mechanism cannot activate.

Hence, I overrode the rendering function (within my application) of PlotView so that it will fire an event after the 1st render cycle, giving the application an opportunity to make initial adjustments.   For GTK this was:

```csharp
		protected override bool OnExposeEvent(EventExpose e)
		{
			var r = base.OnExposeEvent(e);

			// callback on 1st render (post render)
			if (_firstrender && OnInitialRender != null)
				OnInitialRender(this, this);

			this.firstrender = false;
			return r;
		}
```
Rather than overriding the low-level rendering function, perhaps it makes sense to formalize this in the API?

**Comment by julientrflz** [comment]
Does the DesiredSize property is implemented in Oxyplot.WPF .NET 2.1.0-Preview1 version ?
I am actually facing the same issue, with multiples charts that i can't vertically align because some of them have 5 axis, and others 2. What I want is to fix the plot size, not the plot+axis size.
Could you please update me on this please ?

**Comment by VisualMelon** [comment]
@julientrflz it was changed to `DesiredMargin`, containing additional information, e.g. `Width` is split into the `Left` and `Right` properties.

Thanks for spotting this regression, I'll add it to the release notes if I can find them.

https://github.com/oxyplot/oxyplot/pull/1562

**Comment by julientrflz** [comment]
Yesterday I found a temporary solution, which was to add "white axis". It worked, not really properly but It was doing the trick. I have fix my problem using PlotMargins when creating my Plotmodel. Thanks ! 👍 
`var tmp = new PlotModel { Title = title, PlotAreaBorderColor = OxyColors.Black, PlotMargins=new OxyThickness(300,0,0,50)};`
